### ISSUE TITLE
undefined method `attribute_setter' for nil:NilClass

### ISSUE THREAD

**Issue by Draiken** [thoughtbot/shoulda-matchers#904]
Hello!

I'm receiving an error when a validation fails with the `validates_inclusion_of` matcher. If the validation is present, everything is fine, but without it when building the failure message you get this error:

```
NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
     # ./.gem/ruby/2.3.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:405:in `failure_message'
     # ./.gem/ruby/2.3.0/gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:77:in `public_send'
     # ./.gem/ruby/2.3.0/gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:77:in `try!'
     # ./.gem/ruby/2.3.0/gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:63:in `try'
     # ./.gem/ruby/2.3.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:129:in `failure_reason'
     # ./.gem/ruby/2.3.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `block in failure_message'
     # ./.gem/ruby/2.3.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `tap'
     # ./.gem/ruby/2.3.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `failure_message'
     # ./spec/lib/models/address_spec.rb:22:in `block (2 levels) in <module:Models>'
```

My class and spec look close to this:

```
class Thing
  include Virtus.model
  include ActiveModel::Model

  ARRAY = ["value"] 

  attribute :thing, String
end

RSpec.describe Thing do
  it { is_expected.to validate_inclusion_of(:thing).in_array(Thing::ARRAY) }
end
```

I am using active model outside of rails, configuring `Shoulda` to integrate with `RSpec` and `ActiveModel`. Ruby version is 2.3.0 and ActiveModel version is 4.2.5.1

Maybe this is an incompatibility with Virtus that I'm not aware of?

Thanks

**Comment by halilim** [comment]
I'm getting the same error with a regular Active Record association:

```ruby
describe ClassifiedProperty do
  it "should validate that :option is one of :property's options" do
    is_expected.to(
      validate_inclusion_of(:property_option_id)
        .in_array(subject.property.option_ids)
        .allow_nil
    )
  end
end
````
And similarly, it's gone when the validation is in place.

**Comment by joshuarh** [comment]
Encountering the same issue.

**Model:**
```
describe Activity do
  it { should validate_inclusion_of(:type).in_array([ 'video' ]) }
end
```

**Schema**:
```
  create_table "activities", force: :cascade do |t|
    t.string   "name",        null: false
    t.integer  "session_id",  null: false
    t.datetime "created_at",  null: false
    t.datetime "updated_at",  null: false
    t.string   "description"
    t.string   "type",        null: false
  end
```

**Error**:
```
Failures:

  1) Activity should validate that :type is ‹"video"›
     Failure/Error: it { should validate_inclusion_of(:type).in_array(ActivityTypes.all) }
     
     NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:405:in `failure_message'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:17:in `public_send'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:17:in `try!'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:6:in `try'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:129:in `failure_reason'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `block in failure_message'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `tap'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `failure_message'
     # ./spec/models/activity_spec.rb:12:in `block (2 levels) in <top (required)>'
```

**Comment by joshuarh** [comment]
This validation also cannot be used at all if the validation is on an abstract parent:

**Shared Examples**:
```
shared_examples_for 'an activity' do
  it { should validate_inclusion_of(:type).in_array([ 'video' ]) }
end
```

**Spec**:
```
describe Activity::Video do
  it_behaves_like 'an activity'
end
```

**Models**:

```
class Activity < ApplicationRecord
  self.abstract_class = true
  validates :type, inclusion: [ 'video' ]
end
```

```
class Activity::Video < Activity
  self.table_name = 'activities'

  def type
    'video'
  end
end
```

**Error**:

```
Failures:

  1) Activity::Video behaves like an activity should validate that :type is ‹"video"›
     Failure/Error: it { should validate_inclusion_of(:type).in_array([ 'video' ]) }
     
     NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
     Shared Example Group: "an activity" called from ./spec/models/activity/video_spec.rb:2
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:405:in `failure_message'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:17:in `public_send'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:17:in `try!'
     # ./.bundle/ruby/2.2.0/gems/activesupport-5.0.0.beta3/lib/active_support/core_ext/object/try.rb:6:in `try'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:129:in `failure_reason'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `block in failure_message'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `tap'
     # ./.bundle/ruby/2.2.0/gems/shoulda-matchers-3.1.1/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `failure_message'
     # ./spec/models/activity/activity_shared_spec.rb:8:in `block (2 levels) in <top (required)>'
```

**Comment by meoooh** [comment]
I have same issue. I'm using shoulda-matcher 3.1.1

**Comment by jiggneshhgohel** [comment]
I am facing the same error when specifying `with_message` qualifier with a regex containing wildcard character matcher or an interpolation:

```
     NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
```

Details below:

```
class CheckinContact < ActiveRecord::Base
  SUPPORTED_CONTACT_LEVELS = %w( primary secondary )

  belongs_to :user

  validates_presence_of :contact_level, :user
  validates_inclusion_of :contact_level, in: SUPPORTED_CONTACT_LEVELS, message: "is unacceptable. It can be either one of the following: #{SUPPORTED_CONTACT_LEVELS}"
end
```

```
require 'rails_helper'

RSpec.describe CheckinContact, type: :model do
  it { should respond_to(:contact_level) }

  it { should belong_to(:user) }

  it { should validate_presence_of(:contact_level) }
  it { should validate_presence_of(:user) }

  it {
    supported_contact_levels = CheckinContact::SUPPORTED_CONTACT_LEVELS
    should validate_inclusion_of(:contact_level).in_array(supported_contact_levels).with_message(/.* is unacceptable\. It can be either one of the following: #{supported_contact_levels}/)
  }
end
```


```
CheckinContact
  should respond to #contact_level
  should belong to user
  should validate that :contact_level cannot be empty/falsy
  should validate that :user cannot be empty/falsy
  should validate that :contact_level is either ‹"primary"› or ‹"secondary"›, producing a custom validation error on failure (FAILED - 1)

Failures:

  1) CheckinContact should validate that :contact_level is either ‹"primary"› or ‹"secondary"›, producing a custom validation error on failure
     Failure/Error: should validate_inclusion_of(:contact_level).in_array(supported_contact_levels).with_message(/.* is unacceptable\. It can be either one of the following: #{supported_contact_levels}/)
     
     NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
     # .../gems/shoulda-matchers-3.1.0/lib/shoulda/matchers/active_model/allow_value_matcher.rb:405:in `failure_message'
     # .../gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:77:in `public_send'
     # .../gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:77:in `try!'
     # .../gems/activesupport-4.2.5.1/lib/active_support/core_ext/object/try.rb:63:in `try'
     # .../gems/shoulda-matchers-3.1.0/lib/shoulda/matchers/active_model/validation_matcher.rb:129:in `failure_reason'
     # .../gems/shoulda-matchers-3.1.0/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `block in failure_message'
     # .../gems/shoulda-matchers-3.1.0/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `tap'
     # .../gems/shoulda-matchers-3.1.0/lib/shoulda/matchers/active_model/validation_matcher.rb:55:in `failure_message'
     # ./spec/models/checkin_contact_spec.rb:13:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:61:in `block (3 levels) in <top (required)>'
     # .../gems/database_cleaner-1.5.1/lib/database_cleaner/generic/base.rb:16:in `cleaning'
     # .../gems/database_cleaner-1.5.1/lib/database_cleaner/base.rb:92:in `cleaning'
     # .../gems/database_cleaner-1.5.1/lib/database_cleaner/configuration.rb:86:in `block (2 levels) in cleaning'
     # .../gems/database_cleaner-1.5.1/lib/database_cleaner/configuration.rb:87:in `call'
     # ../gems/database_cleaner-1.5.1/lib/database_cleaner/configuration.rb:87:in `cleaning'
     # ./spec/rails_helper.rb:60:in `block (2 levels) in <top (required)>'

```

I have confirmed that using the regex without any wildcard matcher or interpolation the spec passes.

```
  it {
    supported_contact_levels = CheckinContact::SUPPORTED_CONTACT_LEVELS
    should validate_inclusion_of(:contact_level).in_array(supported_contact_levels).with_message(/is unacceptable\. It can be either one of the following/)
  }
```

**Comment by dabrorius** [comment]
I'm having the same issue, did anyone ever figure out what's wrong here?

**Comment by danlaffan** [comment]
+1

**Comment by fcarrega** [comment]
+1

**Comment by asfarto** [comment]
+1

**Comment by mcmire** [comment]
Hey, sorry everyone, I haven't had a chance to look at this yet, as I've been out. Thanks for your patience.

**Comment by Ruff9** [comment]
+1

**Comment by mwenger1** [comment]
👍

**Comment by devmarwen** [comment]
+1

**Comment by phthhieu** [comment]
+1

**Comment by DexterTheDragon** [comment]
Problem appears to come from this commit https://github.com/thoughtbot/shoulda-matchers/commit/296211211497e624dde87adae68b385ad4cdae3a. Specifically, it seems `matches?` and `does_not_match?` of `AllowValueMatcher` can set `@result` to nil, but `failure_message` and `failure_message_when_negated` call `result.attribute_setter` which causes the undefined method on nil error. Hopefully this helps narrow it down some.

**Comment by mcmire** [comment]
I just got this error as well when saying `should_not validate_confirmation_of(:some_attribute)`. It doesn't matter if the test actually passes or not, it will still crash.

**Comment by Mange** [comment]
What's the status on this one? Has anyone come up with a solution?

**Comment by casiodk** [comment]
+1

**Comment by jedi1156** [comment]
+1

**Comment by Zsheng** [comment]
+1
NoMethodError:
       undefined method `validate_inclusion_of' for #<RSpec::ExampleGroups::RetailerRegistry::Validations:0x007fc4a360e978>
       Did you mean?  validate_numericality_of

**Comment by jwillette-sh** [comment]
+1

  1) User attendee should not validate that :email cannot be empty/falsy
     Failure/Error: it { should_not validate_presence_of :email }

     NoMethodError:
       undefined method `attribute_setter' for nil:NilClass
     # ./spec/models/user_spec.rb:27:in `block (3 levels) in <top (required)>'

**Comment by Carlows** [comment]
+1

**Comment by Jay-Plumb** [comment]
+1 `NoMethodError: undefined method `attribute_setter' for nil:NilClass`

**Comment by mcmire** [comment]
This error will also occur when testing the presence of a validation that doesn't exist on the model.

**Comment by ktopolski** [comment]
:+1: this is still happening on inclusion matcher on ActiveRecord object, shoulda 3.1.2, rails 5.1.4

**Comment by ktopolski** [comment]
it can be workarounded with `allow_value` matcher tho:

```ruby
  describe 'my_attr values' do
    it { is_expected.to allow_value('a').for(:my_attr) }
    it { is_expected.to allow_value('b').for(:my_attr) }
    it { is_expected.not_to allow_value('anything other').for(:my_attr) }
  end
```

**Comment by wilianto** [comment]
My investigation result for this issue:

When using `should_not` in RSpec it starts with `NegativeExpectationHandler`. 
Here is the flow:
```
NegativeExpectationHandler#handle_matcher  => 
	ValidatePresenceOfMatcher#matches? =>
	DisallowValueMatcher#matches? =>
->	AllowValueMatcher#does_not_match? =>
	DisallowValueMatcher#failure_message_when_negated => 
->	AllowValueMatcher#failure_message
```

`AllowValueMatcher#does_not_match?` should call `AllowValueMatcher#failure_message_when_negated` instead of `AllowValueMatcher#failure_message`. 

So my suggestion flow is:
```
NegativeExpectationHandler#handle_matcher  => 
	ValidatePresenceOfMatcher#does_not_match? =>
	AllowValueMatcher#match? =>
	AllowValueMatcher#failure_message
```

Any feedback for a new flow? Or I will try to create a PR to implement `does_not_match?` in `ValidatePresenceOfMatcher`.

Thanks

**Comment by mcmire** [comment]
@wilianto 

That's not a bad idea, but we would probably have to make the same change across the board, so I would hesitate to do that.

I'm (slowly) working on a branch that will change how "submatchers" (those are matchers that are used internally by other matches) are processed and should address this issue here and anywhere else it appears.

**Comment by mcmire** [comment]
@guialbuk I just noticed your new 4.0 issue and I just want to bring up this issue because it probably needs to be fixed before 4.0 proper goes out. If people who aren't aware of this new release upgraded cold from what's in master, a bunch of their tests would break I think. I have a branch to fix this but it's basically just exploded at this point and I have no hope of ever finishing it :( I don't know how many people actually use this gem anymore (I know we don't at my company), so it's up to you, but I thought I'd just bring this to your attention.

**Comment by mcmire** [comment]
This should be fixed in master now. We're cooking up an RC so that y'all can try out the fix. We'll post again when we have that.
